File: web/src/api/sync.ts
import { api } from "./client";
import type { components } from "./schema";

export type SyncedPlaylist = components["schemas"]["SyncedPlaylistResponse"];
export type SyncConfig = components["schemas"]["SyncConfigResponse"];
export type SyncStatus = components["schemas"]["SyncStatusResponse"];

export type AddPlaylistResult =
  | { success: true; id: string }
  | { success: false; error: string };

export type SyncResult =
  | { success: true; jobId: string }
  | { success: false; error: string };

export type SyncAllResult =
  | { success: true; jobIds: string[] }
  | { success: false; error: string };

// --- Playlists ---

export async function listPlaylists(): Promise<SyncedPlaylist[]> {
  const { data, error } = await api.GET("/sync/playlists");
  if (error) return [];
  return data.playlists;
}

export async function addPlaylist(
  url: string,
  name: string,
): Promise<AddPlaylistResult> {
  const { data, error, response } = await api.POST("/sync/playlists", {
    body: { url, name },
  });

  if (error) {
    if (response.status === 409) {
      const conflict = error as { message: string };
      return { success: false, error: conflict.message };
    }
    if (response.status === 422) {
      const validation = error as { detail?: { msg: string }[] };
      return {
        success: false,
        error: validation.detail?.[0]?.msg ?? "Invalid input",
      };
    }
    return { success: false, error: "Failed to add playlist" };
  }

  return { success: true, id: data.id };
}

export async function getPlaylist(id: string): Promise<SyncedPlaylist | null> {
  const { data, error } = await api.GET("/sync/playlists/{playlist_id}", {
    params: { path: { playlist_id: id } },
  });
  if (error) return null;
  return data;
}

export async function updatePlaylist(
  id: string,
  updates: { name?: string; enabled?: boolean },
): Promise<SyncedPlaylist | null> {
  const { data, error } = await api.PATCH("/sync/playlists/{playlist_id}", {
    params: { path: { playlist_id: id } },
    body: updates,
  });
  if (error) return null;
  return data;
}

export async function deletePlaylist(id: string): Promise<boolean> {
  const { error } = await api.DELETE("/sync/playlists/{playlist_id}", {
    params: { path: { playlist_id: id } },
  });
  return !error;
}

// --- Sync Jobs ---

export async function syncPlaylist(id: string): Promise<SyncResult> {
  const { data, error, response } = await api.POST(
    "/sync/playlists/{playlist_id}/sync",
    {
      params: { path: { playlist_id: id } },
    },
  );

  if (error) {
    if (response.status === 404) {
      return { success: false, error: "Playlist not found" };
    }
    if (response.status === 409) {
      return { success: false, error: "Job queue is full" };
    }
    return { success: false, error: "Failed to create sync job" };
  }

  return { success: true, jobId: data.job_id };
}

export async function syncAll(): Promise<SyncAllResult> {
  const { data, error } = await api.POST("/sync/run");

  if (error) {
    return { success: false, error: "Failed to create sync jobs" };
  }

  return { success: true, jobIds: data.job_ids };
}

// --- Config ---

export async function getConfig(): Promise<SyncConfig | null> {
  const { data, error } = await api.GET("/sync/config");
  if (error) return null;
  return data;
}

export async function updateConfig(updates: {
  enabled?: boolean;
  interval_minutes?: number;
}): Promise<SyncConfig | null> {
  const { data, error } = await api.PATCH("/sync/config", {
    body: updates,
  });
  if (error) return null;
  return data;
}

// --- Status ---

export async function getStatus(): Promise<SyncStatus | null> {
  const { data, error } = await api.GET("/sync/status");
  if (error) return null;
  return data;
}

File: packages/api/src/yubal_api/api/routes/sync.py
"""Sync API routes for managing playlist syncing."""

import uuid
from datetime import UTC, datetime

from fastapi import APIRouter, status

from yubal_api.api.dependencies import SyncRepositoryDep, SyncSchedulerDep
from yubal_api.api.exceptions import (
    QueueFullError,
    SyncedPlaylistConflictError,
    SyncedPlaylistNotFoundError,
)
from yubal_api.db.models import SyncedPlaylist
from yubal_api.schemas.sync import (
    AddPlaylistRequest,
    AddPlaylistResponse,
    SyncAllResponse,
    SyncConfigResponse,
    SyncedPlaylistResponse,
    SyncedPlaylistsResponse,
    SyncJobResponse,
    SyncStatusResponse,
    UpdatePlaylistRequest,
    UpdateSyncConfigRequest,
)

router = APIRouter(prefix="/sync", tags=["sync"])


# -- Playlist Endpoints --


@router.get("/playlists", response_model=SyncedPlaylistsResponse)
async def list_playlists(repository: SyncRepositoryDep) -> SyncedPlaylistsResponse:
    """List all synced playlists."""
    playlists = repository.list_playlists()
    return SyncedPlaylistsResponse(
        playlists=[
            SyncedPlaylistResponse.model_validate(p, from_attributes=True)
            for p in playlists
        ]
    )


@router.post(
    "/playlists",
    response_model=AddPlaylistResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_playlist(
    request: AddPlaylistRequest,
    repository: SyncRepositoryDep,
) -> AddPlaylistResponse:
    """Add a playlist for syncing.

    Returns 409 Conflict if the URL is already registered.
    """
    # Check for duplicate URL
    existing = repository.get_playlist_by_url(request.url)
    if existing:
        raise SyncedPlaylistConflictError(
            f"Playlist URL already registered as '{existing.name}'"
        )

    playlist = SyncedPlaylist(
        id=str(uuid.uuid4()),
        url=request.url,
        name=request.name,
        created_at=datetime.now(UTC),
    )
    repository.add_playlist(playlist)

    return AddPlaylistResponse(id=playlist.id)


@router.get("/playlists/{playlist_id}", response_model=SyncedPlaylistResponse)
async def get_playlist(
    playlist_id: str,
    repository: SyncRepositoryDep,
) -> SyncedPlaylistResponse:
    """Get a single synced playlist."""
    playlist = repository.get_playlist(playlist_id)
    if not playlist:
        raise SyncedPlaylistNotFoundError(playlist_id)
    return SyncedPlaylistResponse.model_validate(playlist, from_attributes=True)


@router.patch("/playlists/{playlist_id}", response_model=SyncedPlaylistResponse)
async def update_playlist(
    playlist_id: str,
    request: UpdatePlaylistRequest,
    repository: SyncRepositoryDep,
) -> SyncedPlaylistResponse:
    """Update a synced playlist's name or enabled status."""
    updates = request.model_dump(exclude_unset=True)
    if not updates:
        # No updates provided, just return current state
        playlist = repository.get_playlist(playlist_id)
        if not playlist:
            raise SyncedPlaylistNotFoundError(playlist_id)
        return SyncedPlaylistResponse.model_validate(playlist, from_attributes=True)

    playlist = repository.update_playlist(playlist_id, **updates)
    if not playlist:
        raise SyncedPlaylistNotFoundError(playlist_id)
    return SyncedPlaylistResponse.model_validate(playlist, from_attributes=True)


@router.delete("/playlists/{playlist_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_playlist(
    playlist_id: str,
    repository: SyncRepositoryDep,
) -> None:
    """Remove a playlist from syncing."""
    if not repository.delete_playlist(playlist_id):
        raise SyncedPlaylistNotFoundError(playlist_id)


# -- Sync Job Endpoints --


@router.post("/playlists/{playlist_id}/sync", response_model=SyncJobResponse)
async def sync_playlist(
    playlist_id: str,
    scheduler: SyncSchedulerDep,
) -> SyncJobResponse:
    """Create a sync job for a specific playlist.

    Returns 404 if playlist not found, 409 if queue is full.
    """
    job_id = scheduler.sync_playlist(playlist_id)
    if job_id is None:
        # Check if it's a not-found or queue-full situation
        playlist = scheduler._repository.get_playlist(playlist_id)
        if not playlist:
            raise SyncedPlaylistNotFoundError(playlist_id)
        raise QueueFullError()
    return SyncJobResponse(job_id=job_id)


@router.post("/run", response_model=SyncAllResponse)
async def sync_all(scheduler: SyncSchedulerDep) -> SyncAllResponse:
    """Create sync jobs for all enabled playlists."""
    job_ids = scheduler.sync_all()
    count = len(job_ids)
    if count == 0:
        message = "No enabled playlists to sync"
    elif count == 1:
        message = "Created 1 sync job"
    else:
        message = f"Created {count} sync jobs"
    return SyncAllResponse(job_ids=job_ids, message=message)


# -- Config Endpoints --


@router.get("/config", response_model=SyncConfigResponse)
async def get_config(repository: SyncRepositoryDep) -> SyncConfigResponse:
    """Get sync configuration."""
    config = repository.get_config()
    return SyncConfigResponse(
        enabled=config.enabled,
        interval_minutes=config.interval_minutes,
    )


@router.patch("/config", response_model=SyncConfigResponse)
async def update_config(
    request: UpdateSyncConfigRequest,
    repository: SyncRepositoryDep,
) -> SyncConfigResponse:
    """Update sync configuration."""
    updates = request.model_dump(exclude_unset=True)
    config = repository.update_config(**updates)
    return SyncConfigResponse(
        enabled=config.enabled,
        interval_minutes=config.interval_minutes,
    )


# -- Status Endpoint --


@router.get("/status", response_model=SyncStatusResponse)
async def get_status(
    repository: SyncRepositoryDep,
    scheduler: SyncSchedulerDep,
) -> SyncStatusResponse:
    """Get scheduler status and summary."""
    config = repository.get_config()
    playlists = repository.list_playlists()
    enabled_count = sum(1 for p in playlists if p.enabled)

    return SyncStatusResponse(
        scheduler_running=scheduler.is_running,
        config=SyncConfigResponse(
            enabled=config.enabled,
            interval_minutes=config.interval_minutes,
        ),
        playlist_count=len(playlists),
        enabled_playlist_count=enabled_count,
    )
